MatrixSDK/VoIP/MXCallManager.m

```diff
@@ -228,29 +235,80 @@ - (void)placeCallInRoom:(NSString *)roomId withVideo:(BOOL)video
         {
-            // Do a peer to peer, one to one call
-            MXCall *call = [[MXCall alloc] initWithRoomId:roomId andCallManager:self];
-            if (call)
-            {
-                [calls addObject:call];
+            void (^initCall)(NSString *) = ^(NSString *callSignalingRoomId){
+                // Do a peer to peer, one to one call
+                MXCall *call = [[MXCall alloc] initWithRoomId:roomId callSignalingRoomId:callSignalingRoomId andCallManager:self];
+                if (call)
+                {
+                    [self->calls addObject:call];
 
-                [call callWithVideo:video];
+                    [call callWithVideo:video];
 
-                // Broadcast the new outgoing call
-                [[NSNotificationCenter defaultCenter] postNotificationName:kMXCallManagerNewCall object:call userInfo:nil];
+                    // Broadcast the new outgoing call
+                    [[NSNotificationCenter defaultCenter] postNotificationName:kMXCallManagerNewCall object:call userInfo:nil];
 
-                if (success)
+                    if (success)
+                    {
+                        success(call);
+                    }
+                }
+                else
                 {
-                    success(call);
+                    if (failure)
+                    {
+                        failure(nil);
+                    }
                 }
-            }
-            else
+            };
+            
+            if (self.isVirtualRoomsSupported)
             {
-                if (failure)
+                NSString *directUserId = room.directUserId;
+                
+                if (directUserId)
+                {
+                    [self getVirtualUserFrom:directUserId success:^(MXThirdPartyUserInstance * _Nonnull user) {
+                        [self directCallableRoomWithVirtualUser:user.userId
+                                                   nativeRoomId:roomId
+                                                        timeout:kMXCallDirectRoomJoinTimeout
+                                                     completion:^(MXRoom * _Nullable roomWithVirtualUser, NSError * _Nullable error) {
+                            if (error)
+                            {
+                                if (failure)
+                                {
+                                    failure(error);
+                                }
+                            }
+                            else
+                            {
+                                initCall(roomWithVirtualUser.roomId);
+                            }
+                        }];
+                    } failure:^(NSError * _Nullable error) {
+                        if (error)
+                        {
+                            //  there is a real error
+                            if (failure)
+                            {
+                                failure(nil);
```

Comment:
we could forward error as we have one in this case.

Reply:
Ah yes, good catch.